src/main/scala/ir/transforms/CalleePreservedParam.scala

```diff
@@ -0,0 +1,62 @@
+package ir
+package transforms
+
+import cilvisitor.*
+
+object CalleePreservedParam {
+
+  /**
+   * Asusming single-return and parameter form 
+   */
+
+  val preserved = (19 to 29).map(i => s"R$i").toSet
+
+  def isPreservedParam(v: LocalVar) = {
+    // abi assumption
+    v.varName match {
+      case "R29_in" | "R30_in" | "R31_in" | "R29" | "R30" | "R31" | "R29_out" | "R30_out" | "R31_out" => true
+      case o if preserved.contains(o.stripSuffix("_in").stripSuffix("_out")) => true
+      case _ => false
+    }
+  }
+
+  def transform(p: Program) = {
+    val v = MakePreserved()
+    visit_prog(v, p)
+    assert(invariant.correctCalls(p))
+  }
+
+  class MakePreserved extends CILVisitor {
```

Comment:
optionally, this is just a loop over all procedures - it doesn't need to be a visitor.